File: src/Hedwig/Models/FundingTimeSplitUtilization/FundingTimeSplitUtilization.cs
using System.ComponentModel.DataAnnotations;

namespace Hedwig.Models
{
	public class FundingTimeSplitUtilization
	{
		[Required]
		public int Id { get; set; }
		// Duplicative information (on the report) to make client-side logic easier
		[Required]
		public int ReportingPeriodId { get; set; }
		public ReportingPeriod ReportingPeriod { get; protected set; }

		[Required]
		public int ReportId { get; set; }
		public CdcReport Report { get; set; }
		[Required]
		public int FundingSpaceId { get; set; }
		public FundingSpace FundingSpace { get; set; }

		[Required]
		public int FullTimeWeeksUsed { get; set; }
		[Required]
		public int PartTimeWeeksUsed { get; set; }
	}
}

File: src/Hedwig/ClientApp/src/utils/models/fundingTimeSplitUtilization.ts
import {
	FundingTimeSplitUtilization,
	FundingTime,
	FundingTimeSplit,
	ReportingPeriod,
	CdcReport,
	FundingSpace,
} from '../../generated';
import { DeepNonUndefineable } from '../types';

export function sumWeeksUsed(utils: FundingTimeSplitUtilization[], lesserTime: FundingTime) {
	return utils.reduce((acc, util) => {
		switch (lesserTime) {
			case FundingTime.Full:
				return acc + util.fullTimeWeeksUsed;
			case FundingTime.Part:
				return acc + util.partTimeWeeksUsed;
			default:
				throw new Error('Something impossible happened');
		}
	}, 0);
}

export const getSplitUtilizations = (
	report: CdcReport,
	splitTimeFundingSpaces: FundingSpace[],
	reportingPeriodWeeks: number
) => {
	return report.timeSplitUtilizations && report.timeSplitUtilizations.length
		? report.timeSplitUtilizations
		: splitTimeFundingSpaces.map((fundingSpace) =>
				getSplitUtilization(
					fundingSpace.timeSplit as FundingTimeSplit,
					0,
					reportingPeriodWeeks,
					report.reportingPeriod as ReportingPeriod,
					report
				)
		  );
};

export const getSplitUtilization = (
	timeSplit: FundingTimeSplit,
	lesserWeeksUsed: number,
	reportingPeriodWeeks: number,
	reportingPeriod: ReportingPeriod,
	report: CdcReport
): FundingTimeSplitUtilization => {
	const lesserTime =
		timeSplit.fullTimeWeeks < timeSplit.partTimeWeeks ? FundingTime.Full : FundingTime.Part;
	const greaterWeeksUsed = reportingPeriodWeeks - lesserWeeksUsed;
	return {
		id: timeSplit.id,
		fundingSpaceId: timeSplit.fundingSpaceId,
		reportingPeriodId: reportingPeriod.id,
		reportId: report.id,
		fullTimeWeeksUsed: lesserTime === FundingTime.Full ? lesserWeeksUsed : greaterWeeksUsed,
		partTimeWeeksUsed: lesserTime === FundingTime.Full ? greaterWeeksUsed : lesserWeeksUsed,
	};
};
